### PR TITLE
feat(plonky2x): Fuzz testing

### PR DESCRIPTION
**Pull request by Owliie**

### CODE REVIEW

File: casper-finality-proofs/fuzz/README.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+## Seed Corpus
+
+The corpus is located in the `corpus` directory. It should contain files to seed the fuzzer with example data so that it generates better inputs on the fly. To seed the corpus, for example for the compute shuffled index minimal target, run the following command:
+
+```bash
+./seed/compute_shuffled_index.sh minimal
+```
+
+T
```
Comment (Xearty) [inline casper-finality-proofs/fuzz/README.md:19]: ?

File: casper-finality-proofs/fuzz/fuzz_targets/utils/writer.rs [hunk 2]
```diff
@@ -0,0 +1,32 @@
+use std::{
+    error::Error,
+    fs::{create_dir_all, read_dir, File},
+    io::{BufWriter, Write},
+};
+
+const BASE_PATH: &str = "./results";
+
+pub unsafe fn json_write(
```
Comment (Xearty) [inline casper-finality-proofs/fuzz/fuzz_targets/utils/writer.rs:9]: Try to remove this?

File: casper-finality-proofs/fuzz/seed/compute_shuffled_index.sh [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        echo "Directory doesn't exist. Creating directory..."
+        mkdir -p "$corpus_dir"
+    elif [ ! -d "$corpus_dir" ] || [ -z "$(ls -A "$corpus_dir")" ]; then
+        echo "Directory is empty or doesn't exist. Creating files..."
+    else
+        echo "Directory is not empty. Exiting..."
+        exit 0
+    fi
+
+    echo "The full path of the directory is: $(realpath "$corpus_dir")"
```
Comment (Xearty) [inline casper-finality-proofs/fuzz/seed/compute_shuffled_index.sh:33]: Delete this?

**Review by Xearty — APPROVED** [review]

**Review by NikolayKostadinov21 — APPROVED** [review]

File: casper-finality-proofs/fuzz/fuzz_targets/weigh_justification_and_finalization.rs [hunk 4]
```diff
@@ -83,29 +98,44 @@ fuzz_target!(|data: TestData| {
...
     );
 
-    json_write("weigh_justification_and_finalization".to_owned(), value).unwrap();
+    assert!(output_ref.finalized_checkpoint().epoch == output.0.new_finalized_checkpoint.epoch);
+    assert!(output_ref.finalized_checkpoint().root == output.0.new_finalized_checkpoint.root);
+
+    output_ref
+        .justification_bits()
+        .iter()
+        .map(|x| x)
```
Comment (Xearty) [inline casper-finality-proofs/fuzz/fuzz_targets/weigh_justification_and_finalization.rs:134]: Try .copied()

File: casper-finality-proofs/fuzz/scripts/fuzz.sh [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+CORPUS_DIR="$FUZZ_DIR/corpus/$FUZZ_TEST_NAME"
+ADDITIONAL_ARGS=("${@:2}")
+
+# Create directories with sudo if necessary
+sudo mkdir -p "$ARTIFACTS_DIR"
+sudo mkdir -p "$CORPUS_DIR"
+
+# Run the fuzzing command with additional arguments
+RUSTFLAGS="-fsanitize-coverage=0 --cfg fuzzing -Clink-dead-code -Cdebug-assertions -C codegen-units=1" \
+    sudo cargo run \
```
Comment (Xearty) [inline casper-finality-proofs/fuzz/scripts/fuzz.sh:20]: try running it as a regular user

**Review by Xearty — APPROVED** [review]

**Pull request merged**